### PR TITLE
Support Json annotations

### PR DESCRIPTION
**Pull request by kennymckormick**
Support the new json annotation format, which is more scalable than txt files: 

1. You can use the command `python tools/data/anno_txt2json.py annotation.txt` to convert it from txt to json.
2. The dataset can automatically detect json annotation list and load it.
3. When the repo decides to switch from txt to json, just change the default option of output_format in `build_file_list.py`  to 'json'

### CODE REVIEW

File: tools/data/build_file_list.py [hunk 1]
```diff
@@ -147,6 +154,20 @@ def build_list(split):
     return (train_rgb_list, test_rgb_list), (train_flow_list, test_flow_list)
 
 
+def lines2dictlist(lines, format):
```
Comment (dreamerlin) [inline tools/data/build_file_list.py:157]: Docstring

File: tools/data/build_file_list.py [hunk 2]
```diff
@@ -147,6 +154,37 @@ def build_list(split):
...
+    Args:
+        lines (list): List of lines in 'txt' label format.
+            'frame_dir num_frame label' (rawframes + single-label)
+            'frame_dir num_frame label1 label2 ...' (rawframes + multi-label)
+            'filename label' (videos + single-label)
+            'filename label1 label2 ...' (videos + multi-label)
+        format (str): Data format, choices are 'rawframes' and 'videos'.
+
+    Returns:
+        list: list of dictions
```
Comment (dreamerlin) [inline tools/data/build_file_list.py:170]: dictions ?

File: tools/data/build_file_list.py [hunk 3]
```diff
@@ -147,6 +154,37 @@ def build_list(split):
...
+            dict(
+                frame_dir=line[0],
+                total_frames=int(line[1]),
+                label=[int(x) for x in line[2:]]) for line in lines
+        ]
+    elif format == 'videos':
+        data = [
+            dict(filename=line[0], label=[int(x) for x in line[1:]])
+            for line in lines
+        ]
```
Comment (dreamerlin) [inline tools/data/build_file_list.py:184]:
```
else:
    raise ValueError(f'{format} is unsupported')
```

File: tools/data/build_file_list.py (hunk 2 above)
Comment (dreamerlin) [inline tools/data/build_file_list.py:170]: dictions ?

File: tools/data/build_file_list.py (hunk 3 above)
Comment (kennymckormick) [reply tools/data/build_file_list.py:184]: Think no need to do so. Since in argparse we already restrict the choices of format to be rawframes and videos

File: tools/data/build_file_list.py (hunk 2 above)
Comment (kennymckormick) [reply tools/data/build_file_list.py:170]: more descriptions added

File: mmaction/datasets/base.py [hunk 4]
```diff
@@ -3,6 +3,8 @@
 from abc import ABCMeta, abstractmethod
 
 import mmcv
+import torch
+from mmcv import load
```
Comment (dreamerlin) [inline mmaction/datasets/base.py:7]: Just use `mmcv.load`

File: tools/data/anno_txt2json.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    Args:
+        lines (list): List of lines in 'txt' label format.
+            'frame_dir num_frame label' (rawframes + single-label)
+            'frame_dir num_frame label1 label2 ...' (rawframes + multi-label)
+            'filename label' (videos + single-label)
+            'filename label1 label2 ...' (videos + multi-label)
+        format (str): Data format, choices are 'rawframes' and 'videos'.
+
+    Returns:
+        list[diction]: For rawframes format, each diction has keys: frame_dir,
```
Comment (dreamerlin) [inline tools/data/anno_txt2json.py:42]: list[dict]

File: tools/data/anno_txt2json.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
+import argparse
+
+from mmcv import dump
```
Comment (dreamerlin) [inline tools/data/anno_txt2json.py:3]: `import mmcv` and use `mmcv.dump`

File: tools/data/anno_txt2json.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            'if the arg value is None'))
+    args = parser.parse_args()
+
+    return args
+
+
+def lines2dictlist(lines, format):
+    """Convert lines in 'txt' format to dictions in 'json' format. Currently
+    support single-label and multi-label.
+
```
Comment (dreamerlin) [inline tools/data/anno_txt2json.py:32]:
```
Example of a single-label rawframes annotation file:

.. code-block:: txt

    (frame_dir num_frames label)
    some/directory-1 163 1
    some/directory-2 122 1
    some/directory-3 258 2

Example of a multi-label rawframes annotation file:

.. code-block:: txt

    (frame_dir num_frames label1 label2 ...)
    some/directory-1 163 1 3 5
    some/directory-2 122 1 2
    some/directory-3 258 2

Example of a single-label videos annotation file:

.. code-block:: txt

    (filename label)
    some/path/000.mp4 1
    some/path/001.mp4 1
    some/path/002.mp4 2

Example of a multi-label videos annotation file:

.. code-block:: txt

    (filename label1 label2 ...)
    some/path/000.mp4 1 3 5
    some/path/001.mp4 1 4 8
    some/path/002.mp4 2 4 9
```

File: tools/data/anno_txt2json.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+        help=(
+            'the output file name, use annofile.replace(\'.txt\', \'.json\') '
+            'if the arg value is None'))
+    args = parser.parse_args()
+
+    return args
+
+
+def lines2dictlist(lines, format):
+    """Convert lines in 'txt' format to dictions in 'json' format. Currently
```
Comment (dreamerlin) [inline tools/data/anno_txt2json.py:30]: `diction` -> `dictionary`

File: tools/data/anno_txt2json.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
+import argparse
+
+from mmcv import dump
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description='Convert txt anno list to json')
```
Comment (dreamerlin) [inline tools/data/anno_txt2json.py:8]: `anno` -> `annotation`

File: tools/data/anno_txt2json.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+    parser.add_argument(
+        'annofile', type=str, help='the txt annofile to convert')
```
Comment (dreamerlin) [inline tools/data/anno_txt2json.py:10]: `annofile` -> `annotation file`

Comment (kennymckormick) [reply tools/data/anno_txt2json.py]: 'annofile' in comments is replaced as 'annotation file', while the arg name is not changed to keep it short

**Review by dreamerlin — APPROVED** [review]

**Comment by dreamerlin** [comment]
You can remove the "WIP" label when the PR is ready

**Review by innerlee — APPROVED** [review]
